README.md
[![Version](https://img.shields.io/badge/version-v1.1.6-red)](https://github.com/Jeoxs/khanos/releases/latest)
[![Follow on Twitter](https://img.shields.io/badge/twitter-@thejoseaponte-blue.svg)](https://twitter.com/thejoseaponte)
[![Download Khanos](https://img.shields.io/badge/download-khanos-green)](https://github.com/Jeoxs/khanos/releases)

[<p align="center"><img src="https://user-images.githubusercontent.com/1626493/127225979-9abab89f-0adb-468c-bce5-deaeeffce56b.png"
      alt="Get it now on the Releases Page"
      height="200"></p>](https://github.com/Jeoxs/khanos/releases)

# Khanos
> A Flutter Android Application for Kanboard

A native app developed with Flutter framework which connects to the Kanboard API in your hosting or service.

## Install
<p align="center"><a href="https://play.google.com/store/apps/details?id=com.gojaponte.khanos&pcampaignid=pcampaignidMKT-Other-global-all-co-prtnr-py-PartBadge-Mar2515-1" target="_blank"><img alt='Get it on Google Play' src="https://play.google.com/intl/es-419/badges/static/images/badges/en_badge_web_generic.png"
      alt="Get it now on the Releases Page"
      height="100"></a></p>
      
### Or 

Download the latest release from github [Here](https://github.com/Jeoxs/khanos/releases)

## Features

What's all the bells and whistles this project can perform?
* Create/Delete Projects.
* Create/Modify/Delete/Close Tasks.
* Create/Modify/Delete Subtasks
* Mark subtasks as Completed
* Kanban for Projects
* More features coming soon!

## Screnshots

<a href="https://user-images.githubusercontent.com/1626493/125201147-d0440a80-e23b-11eb-906e-3e7bd0153810.jpg"><img src="https://user-images.githubusercontent.com/1626493/125201147-d0440a80-e23b-11eb-906e-3e7bd0153810.jpg" height="300"></img></a>
<a href="https://user-images.githubusercontent.com/1626493/125201148-d0dca100-e23b-11eb-80ab-7e0b8cedc9a2.jpg"><img src="https://user-images.githubusercontent.com/1626493/125201148-d0dca100-e23b-11eb-80ab-7e0b8cedc9a2.jpg" height="300"></img></a>
<a href="https://user-images.githubusercontent.com/1626493/125201149-d1753780-e23b-11eb-8598-da514c78f9f7.jpg"><img src="https://user-images.githubusercontent.com/1626493/125201149-d1753780-e23b-11eb-8598-da514c78f9f7.jpg" height="300"></img></a>
<a href="https://user-images.githubusercontent.com/1626493/125201152-d20dce00-e23b-11eb-84ab-26e2ca45dec8.jpg"><img src="https://user-images.githubusercontent.com/1626493/125201152-d20dce00-e23b-11eb-84ab-26e2ca45dec8.jpg" height="300"></img></a>
<a href="https://user-images.githubusercontent.com/1626493/125201151-d20dce00-e23b-11eb-825c-20b6cf3bbca6.jpg"><img src="https://user-images.githubusercontent.com/1626493/125201151-d20dce00-e23b-11eb-825c-20b6cf3bbca6.jpg" height="300"></img></a>
<a href="https://user-images.githubusercontent.com/1626493/125201150-d1753780-e23b-11eb-81bd-104ec25cc0c0.jpg"><img src="https://user-images.githubusercontent.com/1626493/125201150-d1753780-e23b-11eb-81bd-104ec25cc0c0.jpg" height="300"></img></a>

## Usage
* Fill the credentials and your Kanboard's endpoint in the Login Form. With this, the app will login to your Kanboard Instance.
## Credits
* [LiveLikeCounter - Flutter-Todolist]
* [VeriTechie]
* [murchu27]
* [bobylapointe63]


[LiveLikeCounter - Flutter-Todolist]: https://github.com/LiveLikeCounter/Flutter-Todolist
[VeriTechie]: https://github.com/VeroMoreno
[murchu27]: https://github.com/murchu27
[bobylapointe63]: https://github.com/bobylapointe63


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][Keep a Changelog] and this project adheres to [Semantic Versioning][Semantic Versioning].

## [Released]

## [1.1.6] - 2021-10-30
### Added
- Swimlane Support for Tasks
- My Tasks page
- Note about 2FA on login

## [1.1.5] - 2021-09-05
### Added
- Users can create personal projects.

### Fixed
- Users can access projects and interact depending on their permission | Thanks: @bobylapointe63.

## [1.1.4] - 2021-09-01
### Fixed
- Users can login regardless of their profile type | Thanks: @bobylapointe63.


## [1.1.3] - 2021-08-28
### Fixed
- Fixed the Dark Theme colors on Login Page after logging out | Thanks: @murchu27.

## [1.1.2] - 2021-08-27
### Changed
- The login URL doesn't need the 'jsonrpc.php' anymore. Users can type only the Kanboard URL or specify the 'jsonrpc.php' endpoint if needed. 

### Fixed
- Uncaught login error redirected to login screen without valid credentials | Thanks: @murchu27.

## [1.1.1] - 2021-07-21
### Added
- Added Avatars to Tasks, subtasks, comments and Kanban
- Changed Logo design to comply with Google Play Policies

### Fixed
- Kanban now refresh elements after pop out from task page.

## [1.1.0] - 2021-07-19
### Added
- Changelog Section in Side Menu showing this Changelog
- Added Tutorial Dialog image for Cards

### Changed 
- Shimmer Animation colors for loading Activities

### Fixed
- Empty Overdue Tasks now shows the Empty Icon Image
- Common Framework Bugs and warnings

## [1.0.9] - 2021-07-18
### Added
- Comments section for tasks. User can add comments and remove them
- Added Overdue Tasks page in Home
- Added Activity Page in Home

### Fixed
- Kanban Columns duplicated when moving Items

## [1.0.8] - 2021-07-16
### Added
- Subtasks can be Reordered by long pressing and Drag.

## [1.0.7] - 2021-07-15
### Added
- Close Task Functionality
- Sound Feedback when tapping Cards
- Time Spent for Subtasks

### Fixed
- Cards Can be re-arranged inside column on Kanban
- Bottom padding on Tasks/Subtasks lists.
- Lowered Circle decoration in AppBar to fix overlaping in dark Mode
- Fixed Project Page task List not refreshing when updating the task


## [1.0.6] - 2021-07-13
### Added
- Kanban Functionality. Now you can drag tasks in Kanban View among columns.
- Added About Page
- Added developer's support link.

### Changed
- Wrapped Tasks/subtasks card's titles text to show full title of task, instead of excerpt

### Fixed
- Going back to login when logging in.


## [1.0.5] - 2021-07-11
### Added
- Dark Mode

## [1.0.4] - 2021-07-10
### Added
- Drawer Menu on Home Page with items
    - Support Link to This Github Repo
    - Logout option
- Projects can now be deleted with Slidable feature.
- Changed package Name as standars suggests. Note: You may want to uninstall the old app, as this one will use a new Package name.

## [1.0.3] - 2021-07-10
### Added
- Redirect to Login Screen if login credentials fails, notifying the user with a notification Bar at the bottom.
- Added Spent Time Field for Tasks
- Added "In Progress" Status for Subtasks. Now Subtasks have all the 3 status (undone, progress, completed).
- Added TimeTracking start/end for subtasks when started/completed

### Changed
- Changed task Page loading animation design.

### Fixed 
- Fixed Validation of response data from Server and added handling
- Fixed Color Picker from Task Form
- Fixed DateTime Fields in Task Form, adding a Limit date/time depending on Start/Due date selection
- Home Screen shows projects by User

## [1.0.2] - 2021-07-09
### Added
- Added Validation to API errors 403 Forbidden in Project Page

### Fixed
- Fixed Ending loading Project Page, because of 403 errors from API

## [1.0.1] - 2021-07-09
### Fixed
- Fixed Signatures for App in AndroidManifest & build.gradle.
- Fixed Title name in Main.dart (This title affects the running application in Android Devices)

## [1.0] - 2021-07-09
### Added
- Initial Release.


<!-- ## [Unreleased]

--- -->

---

<!-- Links -->
[Keep a Changelog]: https://keepachangelog.com/
[Semantic Versioning]: https://semver.org/

<!-- Versions -->
[Released]: https://github.com/Jeoxs/khanos/releases

[1.1.5]: https://github.com/Jeoxs/khanos/compare/v1.1.5...v1.1.6
[1.1.5]: https://github.com/Jeoxs/khanos/compare/v1.1.4...v1.1.5
[1.1.4]: https://github.com/Jeoxs/khanos/compare/v1.1.3...v1.1.4
[1.1.3]: https://github.com/Jeoxs/khanos/compare/v1.1.2...v1.1.3
[1.1.2]: https://github.com/Jeoxs/khanos/compare/v1.1.1...v1.1.2
[1.1.1]: https://github.com/Jeoxs/khanos/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/Jeoxs/khanos/compare/v1.0.9...v1.1.0
[1.0.9]: https://github.com/Jeoxs/khanos/compare/v1.0.8...v1.0.9
[1.0.8]: https://github.com/Jeoxs/khanos/compare/v1.0.7...v1.0.8
[1.0.7]: https://github.com/Jeoxs/khanos/compare/v1.0.6...v1.0.7
[1.0.6]: https://github.com/Jeoxs/khanos/compare/v1.0.5...v1.0.6
[1.0.5]: https://github.com/Jeoxs/khanos/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/Jeoxs/khanos/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/Jeoxs/khanos/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/Jeoxs/khanos/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/Jeoxs/khanos/compare/v1.0...v1.0.1
[1.0]: https://github.com/Jeoxs/khanos/releases/v1.0

[Unreleased]: https://github.com/Jeoxs/khanos/compare/v1.0...HEAD


lib/src/pages/home_page.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:khanos/src/pages/activity_page.dart';
import 'package:khanos/src/pages/my_tasks_page.dart';
import 'package:khanos/src/pages/overdue_page.dart';
import 'package:khanos/src/preferences/user_preferences.dart';
import 'package:khanos/src/providers/auth_provider.dart';
import 'package:khanos/src/providers/dark_theme_provider.dart';
import 'package:khanos/src/providers/user_provider.dart';
import 'package:khanos/src/utils/theme_utils.dart';
import 'package:khanos/src/utils/utils.dart';
import 'package:khanos/src/utils/widgets_utils.dart';
import 'package:provider/provider.dart';
import 'package:shimmer/shimmer.dart';
import 'package:khanos/src/providers/column_provider.dart';
import 'package:khanos/src/providers/project_provider.dart';
import 'package:flutter_slidable/flutter_slidable.dart';
import 'package:url_launcher/url_launcher.dart';

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  int _currentIndex = 0;
  final _prefs = new UserPreferences();
  final projectProvider = new ProjectProvider();
  final columnProvider = new ColumnProvider();
  final userProvider = new UserProvider();
  final authProvider = new AuthProvider();
  bool _darkTheme;
  ThemeData currentThemeData;
  Map<String, dynamic> myProfileInfo;

  @override
  void initState() {
    _darkTheme = _prefs.darkTheme;    
    super.initState();
  }

  int projectsAmount;
  @override
  Widget build(BuildContext context) {
    if (_prefs.newInstall == true) {
      Future.delayed(Duration.zero, () => showChangelog(context));
      _prefs.newInstall = false;
    }

    List<Widget> _children = [];
    _children.add(projectList(context));
    _children.add(OverdueTasksPage());
    _children.add(MyTasksPage());
    _children.add(ActivityPage());

    currentThemeData =
        _darkTheme == true ? ThemeData.dark() : ThemeData.light();
    return Scaffold(
      appBar: normalAppBar('Khanos'),
      // body: projectList(context),
      body: _children[_currentIndex],
      drawer: _homeDrawer(),
      floatingActionButton: FloatingActionButton(
        heroTag: "addHero",
        backgroundColor: Colors.blue,
        child: Icon(Icons.add),
        onPressed: () => Navigator.of(context)
            .pushNamed('newProject')
            .then((_) => setState(() {})),
      ),
      bottomNavigationBar: BottomNavigationBar(     
        type: BottomNavigationBarType.fixed,   
        onTap: onTabTapped,
        currentIndex: _currentIndex,
        items: [
          BottomNavigationBarItem(
            icon: new Icon(Icons.home),
            label: 'Home',
          ),
          BottomNavigationBarItem(
            icon: new Icon(Icons.watch_later_outlined),
            label: 'Overdue Tasks',
          ),
          BottomNavigationBarItem(
            icon: new Icon(Icons.check_rounded),
            label: 'My Tasks',
          ),
          BottomNavigationBarItem(
            icon: new Icon(Icons.list),
            label: 'Activities',
          ),
        ],
        selectedItemColor: Colors.blue,
      ),
    );
  }

  Widget homeContent(BuildContext context) {
    return Column(
      children: [
        projectsOverview(context),
        Container(
          padding: EdgeInsets.only(bottom: 5.0, left: 25.0),
          width: double.infinity,
          child: Text(
            'All Projects',
            style: TextStyle(fontWeight: FontWeight.w500, fontSize: 20.0),
            textAlign: TextAlign.left,
          ),
        ),
        Container(child: projectList(context)),
      ],
    );
  }

  Widget projectsOverview(BuildContext context) {
    return Card(
      elevation: 5.0,
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10.0)),
      margin: EdgeInsets.symmetric(vertical: 20.0, horizontal: 20.0),
      child: Padding(
        padding: const EdgeInsets.all(10.0),
        child: Row(
          children: [
            Text(
              '${projectsAmount.toString()} Projects',
              style: TextStyle(fontWeight: FontWeight.w400, fontSize: 25.0),
            ),
          ],
        ),
      ),
    );
  }

  Widget projectList(BuildContext context) {
    currentThemeData =
        _darkTheme == true ? ThemeData.dark() : ThemeData.light();
    return FutureBuilder(
        future:
            Future.wait([projectProvider.getProjects(), authProvider.getMe()]),
        builder: (BuildContext context, AsyncSnapshot snapshot) {
          if (snapshot.hasError) {
            processApiError(snapshot.error);
            if (_prefs.authFlag != true) {
              final SnackBar _snackBar = SnackBar(
                content: const Text('Login Failed!'),
                duration: const Duration(seconds: 5),
              );
              @override
              void run() {
                scheduleMicrotask(() {
                  ScaffoldMessenger.of(context).showSnackBar(_snackBar);
                  Navigator.pushReplacementNamed(context, 'login',
                      arguments: {'error': snapshot.error});
                });
              }

              run();
            } else {
              return Container(
                  width: double.infinity,
                  padding: EdgeInsets.only(top: 20.0),
                  child: errorPage(snapshot.error));
            }
          }
          if (snapshot.hasData) {
            final projects = snapshot.data[0];
            myProfileInfo = snapshot.data[1];
            _prefs.appRole = myProfileInfo['role'];
            projectsAmount = projects.length;
            return Column(
              children: [
                Container(
                  margin: EdgeInsets.only(top: 15, left: 20, bottom: 10),
                  child: Text(
                    'Projects',
                    style: TextStyle(
                      fontSize: 16,
                      fontWeight: FontWeight.w600,
                    ),
                  ),
                ),
                Expanded(
                  child: ListView.builder(
                      padding: EdgeInsets.only(
                          left: 20.0, right: 20.0, bottom: 80.0),
                      itemCount: projects.length,
                      itemBuilder: (BuildContext context, int i) {
                        return GestureDetector(
                          onTap: () {
                            Feedback.forTap(context);
                            Navigator.pushNamed(context, 'project',
                                    arguments: {'project': projects[i]})
                                .then((_) => setState(() {}));
                          },
                          child: Slidable(
                            actionPane: SlidableDrawerActionPane(),
                            child: _projectElement(
                                projects[i].name, projects[i].description, projects[i].isPrivate),
                            secondaryActions: <Widget>[
                              SlideAction(
                                child: Container(
                                  padding: EdgeInsets.only(bottom: 10),
                                  child: Container(
                                    height: 35,
                                    width: 35,
                                    decoration: BoxDecoration(
                                        borderRadius: BorderRadius.circular(50),
                                        color: CustomColors.TrashRedBackground),
                                    child:
                                        Image.asset('assets/images/trash.png'),
                                  ),
                                ),
                                onTap: () {
                                  showLoaderDialog(context);
                                  _removeProject(projects[i].id);
                                },
                              ),
                            ],
                          ),
                        );
                      }),
                ),
              ],
            );
          } else {
            return Column(
              children: [
                Expanded(
                  child: Shimmer.fromColors(
                    child: ListView.builder(
                        padding: EdgeInsets.symmetric(
                            horizontal: 20.0, vertical: 10.0),
                        itemCount: 8,
                        itemBuilder: (context, index) {
                          return _projectElement('someTitle', 'description..',"0");
                        }),
                    baseColor: CustomColors.BlueDark,
                    highlightColor: Colors.lightBlue[200],
                  ),
                ),
              ],
            );
          }
        });
  }

  Widget _projectElement(String title, String description, String isPrivate) {
    description = description != null ? description : 'No description';
    return Container(
      margin: EdgeInsets.symmetric(horizontal: 10.0, vertical: 8.0),
      padding: EdgeInsets.symmetric(vertical: 20.0, horizontal: 30.0),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Column(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: <Widget>[
              Container(
                width: 200,
                child: Text(
                  title,
                  style: TextStyle(fontSize: 20.0),
                  overflow: TextOverflow.ellipsis,
                ),
              ),
              Container(
                width: 200,
                child: Text(description,
                    style: TextStyle(fontSize: 15),
                    overflow: TextOverflow.ellipsis),
              ),
            ],
          ),
        (isPrivate == "1") ? Icon(Icons.lock,color:Colors.grey) : SizedBox(),
        ],
      ),
      decoration: BoxDecoration(
        gradient: LinearGradient(
          stops: [0.015, 0.015],
          colors: [Colors.blue, currentThemeData.cardColor],
        ),
        borderRadius: BorderRadius.all(
          Radius.circular(5.0),
        ),
        boxShadow: [
          BoxShadow(
            color: currentThemeData.shadowColor,
            blurRadius: 4,
            offset: Offset(1.5, 1.5),
          ),
        ],
      ),
    );
  }

  void _removeProject(String projectId) async {
    bool result = await projectProvider.removeProject(int.parse(projectId));
    Navigator.pop(context);
    if (result) {
      setState(() {});
    } else {
      mostrarAlerta(context, 'Something went Wront!');
    }
  }

  Widget _homeDrawer() {
    var _themeProvider = Provider.of<ThemeChanger>(context);
    return Container(
      child: Drawer(
        child: ListView(
          padding: EdgeInsets.zero,
          children: <Widget>[
            Stack(children: [
              DrawerHeader(
                child: Container(),
                decoration: BoxDecoration(
                  gradient: LinearGradient(
                    begin: Alignment.topLeft,
                    end: Alignment.bottomRight,
                    colors: [
                      CustomColors.HeaderBlueDark,
                      CustomColors.HeaderBlueLight
                    ],
                  ),
                  // color: Colors.blueAccent,
                ),
              ),
              CustomPaint(
                painter: CircleOne(),
              ),
              Positioned(
                child: Center(
                  child: Container(
                    margin: EdgeInsets.only(top: 30.0),
                    height: 130.0,
                    width: 130.0,
                    decoration: BoxDecoration(
                      borderRadius: BorderRadius.circular(100),
                      image: DecorationImage(
                          image: AssetImage(
                              'assets/images/khanos_transparent.png')),
                    ),
                  ),
                ),
              ),
              Positioned(
                  child: Center(
                    child: Text('Khanos',
                        style: TextStyle(
                          fontSize: 30.0,
                          color: Colors.white,
                        )),
                  ),
                  top: 140.0,
                  left: 100.0),
            ]),
            SwitchListTile(
              activeColor: Colors.blue,
              value: _darkTheme,
              title: Text('Dark Mode'),
              onChanged: (value) {
                _themeProvider.setTheme(value ? darkTheme : lightTheme);
                prefs.darkTheme = value;
                _darkTheme = value;
                setState(() {});
              },
            ),
            ListTile(
              leading: Icon(Icons.support, color: Colors.blue),
              title: Text('Help & Feedback'),
              onTap: () {
                _launchGithubURL();
              },
            ),
            ListTile(
              leading: Icon(Icons.coffee, color: Colors.blue),
              title: Text('Show your support'),
              onTap: () {
                _launchBuyMeACoffee();
              },
            ),
            ListTile(
              leading: Icon(Icons.bolt, color: Colors.blue),
              title: Text('What\'s New'),
              onTap: () {
                Navigator.pushNamed(context, 'changelog');
              },
            ),
            ListTile(
              leading: Icon(Icons.contact_support_outlined, color: Colors.blue),
              title: Text('About Khanos'),
              onTap: () {
                Navigator.pushNamed(context, 'about');
              },
            ),
            ListTile(
              leading: Icon(Icons.logout, color: Colors.blue),
              title: Text('Logout'),
              onTap: () {
                _prefs.authFlag = false;
                _prefs.userId = 0;
                Navigator.of(context)
                    .pushNamedAndRemoveUntil('login', (route) => false);
              },
            ),
          ],
        ),
      ),
    );
  }

  _launchGithubURL() async {
    print('here');
    const url = 'https://github.com/Jeoxs/khanos';
    if (await canLaunch(url)) {
      await launch(url);
    } else {
      throw 'Could not launch $url';
    }
  }

  _launchBuyMeACoffee() async {
    print('here');
    const url = 'https://buymeacoffee.com/joseaponte';
    if (await canLaunch(url)) {
      await launch(url);
    } else {
      throw 'Could not launch $url';
    }
  }

  void onTabTapped(int index) {
    currentThemeData =
        _darkTheme == true ? ThemeData.dark() : ThemeData.light();
    setState(() {
      _currentIndex = index;
    });
  }
}
